hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/upgrade/ReconLayoutVersionManager.java

```diff
@@ -0,0 +1,146 @@
+ */
+
+package org.apache.hadoop.ozone.recon.upgrade;
+
+import org.apache.hadoop.ozone.recon.ReconSchemaVersionTableManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * ReconLayoutVersionManager is responsible for managing the layout version of the Recon service.
+ * It determines the current Metadata Layout Version (MLV) and Software Layout Version (SLV) of the
+ * Recon service, and finalizes the layout features that need to be upgraded.
+ */
+public class ReconLayoutVersionManager {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ReconLayoutVersionManager.class);
+
+  private final ReconSchemaVersionTableManager schemaVersionTableManager;
+
+  // Metadata Layout Version (MLV) of the Recon Metadata on disk
+  private int currentMLV;
+  // Software Layout Version (SLV) of the Recon service
+  private final int currentSLV;
+
+  public ReconLayoutVersionManager(ReconSchemaVersionTableManager schemaVersionTableManager) {
+    this.schemaVersionTableManager = schemaVersionTableManager;
+    this.currentMLV = determineMLV();
+    this.currentSLV = determineSLV();
+    ReconLayoutFeature.registerUpgradeActions();  // Register actions via annotation
+  }
+
+  /**
+   * Determines the current Metadata Layout Version (MLV) from the version table.
+   * @return The current Metadata Layout Version (MLV).
+   */
+  private int determineMLV() {
+    return schemaVersionTableManager.getCurrentSchemaVersion();
+  }
+
+  /**
+   * Determines the Software Layout Version (SLV) based on the latest feature version.
+   * @return The Software Layout Version (SLV).
+   */
+  private int determineSLV() {
+    return Arrays.stream(ReconLayoutFeature.values())
+        .mapToInt(ReconLayoutFeature::getVersion)
+        .max()
+        .orElse(0); // Default to 0 if no features are defined
+  }
+
+  /**
+   * Finalizes the layout features that need to be upgraded, by executing the upgrade action for each
+   * feature that is registered for finalization.
+   */
+  public void finalizeLayoutFeatures() {
+    // Get features that need finalization, sorted by version
+    List<ReconLayoutFeature> featuresToFinalize = getRegisteredFeatures();
+
+    for (ReconLayoutFeature feature : featuresToFinalize) {
+      try {
+        // Fetch only the AUTO_FINALIZE action for the feature
+        Optional<ReconUpgradeAction> action = feature.getAction(ReconUpgradeAction.UpgradeActionType.AUTO_FINALIZE);
+        if (action.isPresent()) {
+          // Execute the upgrade action & update the schema version in the DB
+          action.get().execute();
+          updateSchemaVersion(feature.getVersion());
+          LOG.info("Feature {} finalized successfully.", feature.getVersion());
+        }
+      } catch (Exception e) {
+        LOG.error("Failed to finalize feature {}: {}", feature.getVersion(), e.getMessage());
```

Comment:
Since after introduction of this framework, features having schema update will be auto-finalized, and in case of any unexpected error or exception, this will silently be eaten up or logged without knowing of error if feature was actually upgraded or not and we as developers will think that feature will be released and upgraded, however this may be logged as silent failure, then its going to be difficult to know with lot of unpleasant surprises. Atleast we need to update this failure with proper details to ReconContext as well as should fail the Recon start, till we have alert mechanism.

Reply:
Will now be tracking failed updates in recon context

Reply:
> Will now be tracking failed updates in recon context

I don't think this behavior is correct. Here we should fail to start Recon along with updating in ReconContext, else we'll never know what happened until user comes back and report us of any issues and our expectation is different.

Reply:
Made changes